### PR TITLE
Adds Teleportation Sickness

### PR DESCRIPTION
**Pull request by BurgerLUA**
https://www.youtube.com/watch?v=x23I8f9PwlI

# WIP PR

I might have missed some things here in terms of oversights/balance/etc.

Exact values, such as how many teleports can trigger the sickness, or how long for each teleport to linger for, can be tweaked to fit balance.

## About The Pull Request

Adds Teleportation Sickness: A new status effect added to those who teleported recently. Teleporting again with this status effect will cause you to vomit, stunning you for a short amount of time. There is no way to negate the vomiting, however having the strong stomach trait will reduce stun times. The status effect lasts 15 seconds, and is cleared when you vomit. This only affects the /mob/living/carbon subtype, which is currently humanoids and xenomorphs. Having the status effect itself does not draw any penalties; just teleporting again with the status effect on you will incur penalties.

This does not affect wormhole teleportation.

Ninjas are immune to this effect since that is pretty much their only source of mobility.

## Why It's Good For The Game

Teleportation is currently in an unbalanced state right now where it is entirely possible to be untouchable by exploiting many teleportation mechanics at once. Most teleportation mechanics have their own cooldown and/or way of limiting repeat use, but there is nothing stopping someone from using two or even three teleportation mechanics at once to become untouchable.

This PR solves that by basically adding a shared consequence for using multiple teleportation methods in a short amount of time.

## Proof Of Testing

99% tested. Still needs some other things to do.

## TODO

- [x] Test interaction with the anti-teleportation sickness trait.
- [x] Test interaction with wormholes.
- [x] Add a better status alert icon for having teleportation sickness.
- [x] Add a cooldown notification like how food has (Maybe? IDK).
~~- [ ] Add variation in teleportation sickness based on distance traveled.~~ Not Feasible
- [ ] Testmerge.
- [ ] Balance per testmerge feedback.

## Changelog

:cl: BurgerBB
balance: Adds Teleportation Sickness: A new status effect added to those who teleported recently. Teleporting again with this status effect will cause you to vomit, stunning you for a short amount of time.
/:cl:

### CODE REVIEW

**Comment by pixelkitty286** [comment]
You are actively nerfing a thing that is a core mechanic of the game please reconsider.

**Comment by BurgerLUA** [comment]
> You are actively nerfing a thing that is a core mechanic of the game please reconsider.

Teleportation, while a very useful mechanic, is not as core to warrant complete immunity from balance changes.

Teleportation is not getting removed, but being balanced so that excess use of it is not possible.

**Comment by pixelkitty286** [comment]
> > You are actively nerfing a thing that is a core mechanic of the game please reconsider.
> 
> Teleportation, while a very useful mechanic, is not as core to warrant complete immunity from balance changes.
> 
> Teleportation is not getting removed, but being balanced so that excess use of it is not possible.

I mean i see the want but we shouldn't just make it two times or some shit maybe excessive teleporting? after all if you teleport with blue space item you do get obliterated or get teleported to some where random.

**Comment by BurgerLUA** [comment]
If people feel that it is too oppressive, I can add an optimization check to it

``!prob(80)``

**Comment by coffeecoffeecoffee** [comment]
There are many other antagonists that use this as a core mechanic, such as wizard and heretic. What's more, I know of several people whose characters are 'experienced teleporters' (often blueshields) so it could do with an expensive quirk you can take to negate the vomiting as an 'experienced teleporter'.

**Comment by SovietSpaghetti** [comment]
> There are many other antagonists that use this as a core mechanic, such as wizard and heretic. What's more, I know of several people whose characters are 'experienced teleporters' (often blueshields) so it could do with an expensive quirk you can take to negate the vomiting as an 'experienced teleporter'.

neither wizards or heretics should be teleporting more than once every fifteen seconds, nor should anyone, really. 

although this will suck if it affects teleports using wormholes/hand tele/portal guns. mainly because i really like running in a portal loop

**Comment by BurgerLUA** [comment]
Wormholes are affected, hand tele and portal gun are affected.

Tomorrow I will be adding safety effects for portal loops so that an infinite stun isn't possible with this mechanic.

**Comment by nevimer** [comment]
Probably an OK idea.

**Comment by Majkl-J** [comment]
I wonder if you could abuse this by throwing bluespace tomatoes at sec to knock them out of a fight

**Comment by BurgerLUA** [comment]
> I wonder if you could abuse this by throwing bluespace tomatoes at sec to knock them out of a fight

You'd have to hit the same person with two bluespace tomatos within 15 seconds. I'd think that this would be funny to keep considering the skill level required to hit someone twice in a row (they know now what you're doing after the first throw).

This also means that you can actually counter teleporting antagonists by throwing bluespace tomatos at them. Seems pretty cool, actually.

**Comment by Majkl-J** [comment]
> This also means that you can actually counter teleporting antagonists by throwing bluespace tomatos at them. Seems pretty cool, actually.

I kinda wanna see that in action so let's give it a shot with a TM

**Comment by pixelkitty286** [comment]
> Wormholes are affected, hand tele and portal gun are affected.

I feel like this is a bad idea

**Comment by StrangeWeirdKitten** [comment]
Putting this in the teleport signal has a lot of implications. Will need discussion.

**Comment by Arturlang** [comment]
Reduce the time it lasts to 10 seconds or less

**Comment by KazooBard** [comment]
> > There are many other antagonists that use this as a core mechanic, such as wizard and heretic. What's more, I know of several people whose characters are 'experienced teleporters' (often blueshields) so it could do with an expensive quirk you can take to negate the vomiting as an 'experienced teleporter'.
> 
> neither wizards or heretics should be teleporting more than once every fifteen seconds, nor should anyone, really. I feel like people are vastly exaggerating how punishing this will be when the time you can't teleport without vomiting is _fifteen seconds_
> 
> although this will suck if it affects teleports using wormholes/hand tele/portal guns. mainly because i really like running in a portal loop

The heretic's cooldown on jaunt is 10 seconds and is a method to break in and break out. It's designed to teleport faster than once every 15 seconds

**Comment by pixelkitty286** [comment]
i just experienced this for my self this should never get merged 1 warning then a hard stun what the fuck this makes everything that is for teleporting impossible now. this is almost as bad as 2 flash stun on borgs but that has a reason to exist

**Comment by StrangeWeirdKitten** [comment]
Alright, I did a bit of discussion and it's had a little bit of time getting tested on the server. 
- Putting it inside the core teleport signal has a lot of implications.
- A lot of teleportation in regards to antags is mostly a practice of being quick, and the distance. A heretic shattering their blade vs a wizard jaunting is a much broader teleport and escape method.
- It punishes the station more then it does the actual antags. It makes the hand teleporter, or any brief teleportation station side, extremely punishing.

I do agree teleportation is a game of frustration, but this is not the method to go about it. Instead I'm suggesting much better alternatives for security to fight teleportation. We already have grounding implants, there could also be a grounding ammo (which embeds for a short time) or a special item (a use for the dimensional anamoly core perhaps). But overall, the TM is getting pulled.

**Review by StrangeWeirdKitten — CHANGES_REQUESTED** [review]
Alright, I did a bit of discussion and it's had a little bit of time getting tested on the server. 
- Putting it inside the core teleport signal has a lot of implications.
- A lot of teleportation in regards to antags is mostly a practice of being quick, and the distance. A heretic shattering their blade vs a wizard jaunting is a much broader teleport and escape method.
- It punishes the station more then it does the actual antags. It makes the hand teleporter, or any brief teleportation station side, extremely punishing.

I do agree teleportation is a game of frustration, but this is not the method to go about it. Instead I'm suggesting much better alternatives for security to fight teleportation. We already have grounding implants, there could also be a grounding ammo (which embeds for a short time) or a special item (a use for the dimensional anamoly core perhaps). But overall, the TM is getting pulled.

**Comment by theselfish** [comment]
Oh, and as stated yesterday; this breaks Quantum Pads, meaning transport networks become annoying at best and outclassed by walking at worst. They already have in-built cooldowns and...If someone is using the pads for antagging they're either leaving a big opening for Sec to jump in after them or are extremely fast with the screwdriver-crowbar dismantling, which would make it a one use teleport anyway.

**Comment by BurgerLUA** [comment]
> * Putting it inside the core teleport signal has a lot of implications. It causes more issues then solving the few. If multiple things have to be balanced around it, then it's the core mechanic itself.

Not sure what " has a lot of implications" even means. You don't have to balance anything around this PR because if you're designing an antagonist type that spam teleports, then you're not designing an antagonist well.

Worst case scenario, you add the new trait that was given with the PR if there is an antagonist who has to do this, like the ninja. It takes 2 seconds to add.

> * A lot of teleportation in regards to antags is mostly a practice of being quick, and the distance. A heretic shattering their blade vs a wizard jaunting is a much broader teleport and escape method.

If you read the PR description, this is about solving exploiting multiple teleportation mechanics at once, not about stoping Joe McHeretic from breaking their blade.

> * It punishes the station more then it does the actual antags. It makes the hand teleporter, or any brief teleportation station side, extremely punishing.

Quantum teleports (pads, inverters) are no longer affected by this. Wormholes were never affected by this from the start. The only thing this affects is the teleportation station and the hand teleporter, both of which should generally not be used frequently as it leads to balance issues.

I recall rounds on Citadel where there was a captain so good with the hand tele that they could survive nuke ops solo for like 30 minutes. I think it was Kevinz, actually.

**Comment by BurgerLUA** [comment]
> Oh, and as stated yesterday; this breaks Quantum Pads, meaning transport networks become annoying at best and outclassed by walking at worst. They already have in-built cooldowns and...If someone is using the pads for antagging they're either leaving a big opening for Sec to jump in after them or are extremely fast with the screwdriver-crowbar dismantling, which would make it a one use teleport anyway.

Quantum pads were given immunity from this effect yesterday.

**Comment by BurgerLUA** [comment]
Like I don't even know what you want me to do dude. Usually when you "request changes" you list changes, not just say "I don't like this" and give reasons why that I can't work with because you didn't read the PR description.

I can:
- Add randomness to the effect.
- Make it scale with distance (requires non-modular edits).
- Increase or decrease the cooldown.
- Allow more types of teleport not to be affected by the sickness.

**Comment by StrangeWeirdKitten** [comment]
Like I said, the majority of teleports that people use are primarily the long range jaunts like a heretic breaking their blade where this will remain quite ineffective. But another common item is the Hierophrat wand. So I am not really sure what this method to combat teleportation is achieving outside of one or two specific items to warrent having the signal itself trigger teleport sickness. Especially if the cooldown was placed to 10 seconds so antags can still use their abilities like the heretic jaunt. Rapid back to back teleportation *isn't* the issue. It's the ability to instantly vanish, sometimes accross the entire station, Every teleport method already has the cooldowns needed to help manage what this is trying to do. 

Like I said, it punished the station more for what it was trying to achieve. You can add tags to make certain teleports not make you sick, sure. But then what is it doing when everything already has its own cooldowns? Why not directly edit the cooldowns if they are too short?

This is an overcomplication to tackle a couple specific items because it's still not going to stop someone with a quantam inverter, red or blue cube, or hierophrat rod from vanishing into lavaland or teleporting through a window to a room security doesn't have access to. It's why I suggested grounding methods, because that would be far more effective as something to deploy and we then don't have to add arbitary timers when teleport cooldowns already exist to stop the issue this PR is trying to fix.

**Comment by BurgerLUA** [comment]
I don't even know how to respond to you anymore.

Teleportation mechanics an issue. Fix it in another PR, go ahead. This is out of the scope of the PR. This solution isn't overcomplicated in the slightest and I don't know how to even dumb it down further than "Don't teleport more than once in 10 seconds."

You suggest other things to tackle the problems, but you're not implementing them.

**Comment by StrangeWeirdKitten** [comment]
If it's out of scope, then that's your choice I suppose. I don't think there's nothing much else to add. Everything has cooldowns already and this isn't fixing anything.
